Skip merging projects with multiplexed libraries

allyhawkins opened the pull request:
Closes #716 
Closes #717 

Here I added a line to filter out any projects that have multiplexed libraries from the channel that is used as input to create merged objects. 

Because we have decided against including multiplexed libraries being included in merged objects, I removed any support we had for merging multiplexed libraries in the `merge_sces.R` script. I was a little torn on if we should keep this, but if the plan is to eventually create merged objects for all non-multiplexed libraries in a project, then we don't need this anymore. I can add it back if we want the flexibility to use it in the future, but I think a better change to that script would be to spit out an error if any libraries being merged contained multiplexed data when we address #718. 

As part of that, I also removed the columns to keep referring to cell hash stats and demuxing results, resolving the issue in #717.

jashapiro commented on bin/merge_sces.R:
I would probably leave this script alone? Having options that we are not currently using in production does not seem a problem to me, and it might be useful as a base for later testing.

jashapiro commented on merge.nf:
```diff
@@ -162,6 +161,9 @@ workflow {
       .filter{it.seq_unit in ['cell', 'nucleus']}
+      // remove any multiplexed projects
+      // future todo: only filter library ids that are multiplexed, but keep all other non-multiplexed libraries
+      .filter{!(it.project_id in multiplex_projects.getVal())}
```
To facilitate the later implementation, why don't we use `.branch` here and then we can log a message about skipping multiplexed libraries in this report.

jashapiro commented on merge.nf:
```diff
@@ -162,6 +161,9 @@ workflow {
     grouped_libraries_ch = libraries_ch
       // only include single-cell/single-nuclei which ensures we don't try to merge libraries from spatial or bulk data
       .filter{it.seq_unit in ['cell', 'nucleus']}
+      // remove any multiplexed projects
+      // future todo: only filter library ids that are multiplexed, but keep all other non-multiplexed libraries
+      .filter{!(it.project_id in multiplex_projects.getVal())}
       // create tuple of [project id, library_id, processed_sce_file]
       .map{[
```
To facilitate the later implementation, why don't we use .branch here and then we can log a message about skipping multiplexed libraries in this report.

(I used `log.warn`, but maybe just `log.info`?)

```suggestion
    filtered_libraries_ch = libraries_ch
      // only include single-cell/single-nuclei which ensures we don't try to merge libraries from spatial or bulk data
      .filter{it.seq_unit in ['cell', 'nucleus']}
      // remove any multiplexed projects
      // future todo: only filter library ids that are multiplexed, but keep all other non-multiplexed libraries
      .branch{
        multiplexed: it.project_id in multiplex_projects.getVal()
        single_sample: true
      }
    
    filtered_libraries_ch.multiplexed
      .unique{ it.project_id }
      .subscribe{
        log.warn("Not merging ${it.project_id} because it contains multiplexed libraries.")
      }
      
    grouped_libraries_ch = filtered_libraries.single_sample 
      // create tuple of [project id, library_id, processed_sce_file]
      .map{[
```

jashapiro reviewed (commented):
Looks good. I made a couple suggestions to make the future easier, one of which was just to leave the script alone.

allyhawkins commented:
I updated this to use the branching you suggested and then also removed the changes to the script, so that should be the same now, other than addressing the capitalization of `ID` for the demuxing columns. 

I am testing this with running one multiplexed and one non-multiplexed libraries, and I think I like the `log.warn`? It comes up as a message pretty immediately when running so you can see why not as many jobs started as you might expect if a multiplexed library is there: 
<img width="454" alt="Screenshot 2024-03-07 at 9 01 30 AM" src="https://github.com/AlexsLemonade/scpca-nf/assets/54039191/5660888f-35a7-4a89-977a-e295226d8a5c">

jashapiro reviewed (approved):
LGTM!
